doc/portfile.7
.\" portfile.7
.\"
.\" Copyright (c) 2004-2015, 2017-2018 The MacPorts Project
.\" Copyright (c) 2002-2003 Apple Inc.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of Apple Inc. nor the names of its
.\"    contributors may be used to endorse or promote products derived from
.\"    this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd February 13, 2007
.Dt PORTFILE 7 "MacPorts"
.Os Darwin
.Sh NAME
.Nm Portfile
.Nd MacPorts description file reference
.Sh DESCRIPTION
A complete reference of all available
.Nm
variables and example syntax.
.Nm Portfiles
consist of valid TCL, allowing for simple key/value pair syntax as well
as utilization of TCL's extensive functionality.
.Nm Portfiles
are encoded in UTF-8.
.Pp
The
.Nm MacPorts System
uses a target dependency system based on a depends/provides model, allowing
for targets to be registered and executed in the correct order based on their
individual requirements.
.Pp
A
.Nm
author needs to be aware of the various standard targets, the options
that they require and the variables that both the targets and the port
system provide.
.Sh PORTSYSTEM
Portfiles must begin with a PortSystem line that defines which version of the
Portfile interpreter should be used.
.br
.Sy Synopsis:
.Dl PortSystem 1.0
.Sh MAIN VARIABLES
All ports are required to set certain variables.
.Bl -tag -width lc
.It Ic name
Full name of port.
.br
.Sy Type:
.Em required
.br
.Sy Example:
.Dl name XFree86
.It Ic version
Upstream version of software.
.br
.Sy Type:
.Em required
.br
.Sy Example:
.Dl version 4.2.1
.It Ic epoch
If a port's version numbering changes such that a newer version looks
older than the previous version, the
.Ic epoch
should be increased.
Some
.Nm Portfile
authors have used large
.Ic epoch
values that look like a date, but there is no reason to do so.
The
.Ic epoch
is simply an unsigned integer, and the only requirement is that it never be
decreased.
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em 0
.br
.Sy Example:
.Dl epoch 1
.It Ic description
One-line description of the software and what it does.
To appear in the description, brackets and semi-colons need to be escaped with
a backslash (i.e., they must be written as
.Ql \e[ ,
.Ql \e]
and
.Ql \e; ) .
The escape sequences listed in
.Xr re_syntax n
may also be used, with the exception of
.Ql \en ,
.Ql \er
and
.Ql \ef .
.br
.Sy Type:
.Em required
.br
.Sy Example:
.Dl description IMAP email checker
.It Ic long_description
A verbose description of the software and what it does.
To appear in the description, brackets and semi-colons need to be escaped with
a backslash (i.e., they must be written as
.Ql \e[ ,
.Ql \e]
and
.Ql \e; ) .
The escape sequences listed in
.Xr re_syntax n
may also be used, with the exception of
.Ql \en ,
.Ql \er
and
.Ql \ef .
.br
.Sy Type:
.Em required
.br
.Sy Example:
.Bd -literal -offset indent -compact
long_description The GNU Image Manipulation \e
    Program (GIMP) is a powerful tool for the \e
    preparation and manipulation of digital \e
    images. The GIMP provides the user with a \e
    wide variety of image manipulation, painting, \e
    processing, and rendering tools.
.Ed
.It Ic notes
Notes for setup and use of the port.
This is shown after the port is activated and anytime the
.Ic notes
command is used; for example:
.Pp
.Dl "port notes python26"
.Pp
The advantage to using
.Ic notes
instead of
.Ic ui_msg
is that it can be queried after a port is installed but
.Ic ui_msg
is only printed during an install.
Therefore
.Ic notes
is good for any information which may be needed anytime after an install.
.br
.Sy Type:
.Em optional
.br
.Sy Example:
.Bd -literal -offset indent -compact
notes "
To fully complete your installation and make\e
python ${branch} the default, please run:

  sudo port install python_select
  sudo python_select ${name}
"
.Ed
.It Ic revision
Local revision number of
.Nm .
Increment for port revisions which would change its installation in any way.
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em 0
.br
.Sy Example:
.Dl revision 1
.It Ic categories
Categories to which this port belongs.
.br
.Sy Type:
.Em required
.br
.Sy Example:
.Dl categories spelling textproc
.It Ic maintainers
E-mail address(es) of port maintainer(s).
.br
.Sy Type:
.Em required
.br
.Sy Example:
.Dl maintainers landonf@macports.org
.It Ic platforms
Declares which platforms are supported by the port.
.br
.Sy Type:
.Em optional
.br
.Sy Values:
.Em darwin freebsd
.br
.Sy Default:
.Em darwin
.br
.Sy Example:
.Dl platforms darwin freebsd
.It Ic supported_archs
The CPU architectures for which this port can be built.
Archs currently supported by macOS are:
.Em i386 ,
.Em ppc ,
.Em ppc64 ,
and
.Em x86_64 .
If this option is not set, it is assumed that the port can build for all
archs.
If a port does not install any architecture-specific files, use the special
value noarch.
.br
If the building architecture isn't among
.Ic supported_archs ,
port fails with an error message except when building on
.Em x86_64
(or
.Em ppc64 )
and
.Ic supported_archs
contains
.Em i386
(or
.Em ppc ) .
In this case, the port will be built in 32-bit mode.
.br
.Sy Type:
.Em optional
.br
.Sy Values:
.Em i386 x86_64 ppc ppc64 noarch
.br
.Sy Default:
.Em i386 x86_64 ppc ppc64
.br
.Sy Examples:
.Dl supported_archs i386 x86_64
.Dl supported_archs noarch
.It Ic homepage
Project homepage for the port.
.br
.Sy Type:
.Em required
.br
.Sy Example:
.Dl http://wireshark.org
.It Ic master_sites
List of sites to fetch
.Ic distfiles
from or a predefined mirror site list.
If set to a predefined mirror site, without a subdirectory being defined, the
portname is used as the name of the subdirectory.
.br
.Sy Type:
.Em required
.br
.Sy Example:
.Bd -literal -offset indent -compact
master_sites ftp://ftp.cdrom.com/pub/magic \e
    sourceforge
.Ed
.It Ic worksrcdir
Path to source directory relative to
.Ic workpath .
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em ${distname}
.br
.Sy Example:
.Dl worksrcdir ${distname}-src-${version}
.It Ic distname
Name of distribution file, without the
.Cm extract.suffix .
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em ${name}-${version}
.br
.Sy Example:
.Dl distname ${name}-${version}-src
.It Xo
.Ic checksums Ar filename Ar type Ar checksum
.Op Ar filename Ar type checksum ...
.Xc
List of checksums for the
.Ic distfiles .
The checksum
.Ar type
can currently be md5, sha1, rmd160, sha256 or size.
The
.Ar filename
can be omitted if there is only one distfile.
.br
.Sy Type:
.Em required
.br
.Sy Example:
.Bd -literal -offset indent -compact
checksums dictd-1.7.1.tar.gz md5 81317b86ea0a5df0163900ad2e6bb12c \e
          magic-words-1.7.1.tar.gz md5 897a005182928613eadd30c267ce9c5b
.Ed
.br
.Sy Example (ledit 1.11):
.Bd -literal -offset indent -compact
checksums md5 a2d38ba641682509c1e964ad699a9dd2 \e
          sha1 1fb6443b5fdf3c83787953f06282d256477c1288
.Ed
.br
.Sy Example (ssldump 0.9b3):
.Bd -literal -offset indent -compact
checksums md5 ac8c28fe87508d6bfb06344ec496b1dd \e
          sha1 a633a9a811a138eac5ed440d583473b644135ef5 \e
          rmd160 941cf8f2ef8459ec4f9ce65772e134505d46566
.Ed
.It Ic macosx_deployment_target
Value for
.Ev MACOSX_DEPLOYMENT_TARGET
environment variable when invoking the configure script.
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em (current OS version)
.br
.Sy Example:
.Dl macosx_deployment_target 10.8
.It Ic use_parallel_build
If set to yes (and the user has enabled buildmakejobs in
.Pa macports.conf ) ,
the port can be built using more than one job.
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em yes
.br
.Sy Example:
.Dl use_parallel_build no
.It Ic use_automake
If set to yes, run the
.Cm automake
target to build any
.Pa Makefile.in
files for use by
.Pa configure .
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em no
.br
.Sy Example:
.Dl use_automake yes
.It Ic use_autoreconf
If set to yes, run the
.Cm autoreconf
target to generate GNU build system files.
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em no
.br
.Sy Example:
.Dl use_autoreconf yes
.It Ic use_autoconf
If set to yes, run the
.Cm autoconf
target to build any
.Pa configure
script required.
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em no
.br
.Sy Example:
.Dl use_autoconf yes
.It Ic use_configure
If set to yes, run the
.Cm configure
target to configure the build.
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em yes
.br
.Sy Example:
.Dl use_configure no
.It Ic copy_log_files Ar path/to/logfile1 Ar path/to/logfile2 Ar ...
Copy specific log files from the workdir to the main MacPorts log dir.
.br
.Sy Type:
.Em optional
.br
.Sy Example:
.Dl copy_log_files config.log
.It Ic conflicts
Used to list ports which conflict with the one containing the
.Cm conflicts
declaration.
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em none (empty)
.br
.Sy Example:
.Dl conflicts cdrtools
.It Ic replaced_by
When a particular port is deprecated in favor of another, use
.Cm replaced_by
in the deprecated port and list the new one to be used.
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em none (empty)
.br
.Sy Example:
.Dl replaced_by xorg-renderproto
.It Ic add_users
Consists of a list of usernames and settings.
At appropriate times during the port installation process, a user will be
created for each username with the corresponding settings.
Settings are of the form name=value.
A setting applies to the username that appeared most recently before it in the
list.
.Pp
Applicable options are: group, gid (may be used instead of group), passwd,
realname, home, and shell.
.Pp
.Sy Type:
.Em optional
.br
.Sy Default:
.Em none (empty)
.br
.Sy Examples:
.Bd -literal -offset indent -compact
add_users squid group=squid realname=Squid\e Proxy \e
    home=${prefix}/var/squid
add_users user1 group=mygroup user2 group=mygroup
.Ed
.It Ic installs_libs
By default, it is assumed that ports may install libraries or headers that can
be incorporated into their dependents.
If this is not the case, set
.Cm installs_libs
to
.Em no .
This means that this port's dependents need not check that it is installed
for the same architectures as them; that it is permissible to distribute
binaries of the dependents even if their licenses conflict with the license
of this port; and that updates to this port can never result in broken dynamic
linking in its dependents.
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em none
.br
.Sy Example:
.Dl installs_libs no
.It Ic license_noconflict
By default, it is assumed that ports may use libraries or headers from their
dependencies and thus form a derivative work.
A dependency with an incompatible license thus prevents the port from being
distributed in binary form.
If a dependency with an incompatible license is not used in such a way that a
derivative work is formed, or should not prevent binary distribution for any
other reason, add its name to this list.
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em none
.br
.Sy Example:
.Dl license_noconflict openssl
.Dl license_noconflict readline gdbm
.El
.Sh TARGET HOOKS
A number of hooks are available for customizing many of the standard
targets that
.Xr port 1
executes.
The targets supporting these hooks are
.Cm fetch ,
.Cm automake ,
.Cm autoconf ,
.Cm configure ,
.Cm build ,
.Cm destroot ,
and
.Cm test .
The hooks are:
.Bl -tag -width lc
.It Va target Ns Ic .asroot
Run the
.Va target
with root privileges.
.br
.Sy Example:
.Dl install.asroot yes
.It Va target Ns Ic .dir
Directory in which to run the
.Va target .
.br
.Sy Example:
.Dl automake.dir src
.It Va target Ns Ic .env
Change the environment the
.Va target
is run in.
This is often overridden on a per
.Nm
basis.
.br
.Sy Example:
.Bd -literal -offset indent -compact
build.env CC="${configure.cc} [get_canonical_archflags]"
.Ed
.It Va target Ns Ic .pre_args
Additional arguments passed before the main arguments.
.br
.Sy Example:
.Dl extract.pre_args -cd
.It Va target Ns Ic .args
Main arguments to pass to the
.Va target .
This is often overridden on a per
.Nm
basis.
.br
.Sy Example:
.Dl configure.args --enable-fooble
.It Va target Ns Ic .post_args
Additional arguments passed after the main arguments.
.br
.Sy Example:
.Dl extract.post_args | tar xf -
.El
.Sh RUNTIME VARIABLES
Read-only access to the MacPorts configuration is provided.
.Bl -tag -width lc
.It Ic prefix
Install prefix.
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em /opt/local
.It Ic libpath
Location of ports-specific TCL libraries.
.br
.Sy Type:
.Em read-only
.It Ic portpath
Full path to the Portfile location.
.br
.Sy Type:
.Em read-only
.br
.Sy Default:
.Em work
.It Ic workpath
Full path to work directory.
.br
.Sy Type:
.Em read-only
.br
.Sy Default:
.Em ${portbuildpath}/work
.It Ic worksrcpath
Full path to working sources (where port has unpacked itself).
.br
.Sy Type:
.Em read-only
.br
.Sy Default:
.Em ${workpath}/${worksrcdir}
.It Ic filesdir
Path to port files relative to
.Ic portpath .
.br
.Sy Type:
.Em read-only
.br
.Sy Default:
.Em files
.It Ic filespath
Full path to the port files location.
.br
.Sy Type:
.Em read-only
.br
.Sy Default:
.Em ${portpath}/${filesdir}
.It Ic distpath
Location to store downloaded distfiles.
.br
.Sy Type:
.Em read-only
.br
.Sy Default:
.Em ${portdbpath}/distfiles/${dist_subdir}/
.It Ic os.arch
Identifies hardware type (e.g.,
.Dq powerpc ) .
.br
.Sy Type:
.Em read-only
.It Ic os.version
Version number of operating system (e.g.,
.Dq 7.0 ) .
.br
.Sy Type:
.Em read-only
.It Ic os.major
Major version number of operating system (e.g.,
.Dq 7 ) .
.br
.Sy Type:
.Em read-only
.It Ic os.endian
Endianness of the processor (e.g.,
.Dq big ) .
.br
.Sy Type:
.Em read-only
.It Ic os.platform
Operating system name (e.g.,
.Dq darwin ) .
.br
.Sy Type:
.Em read-only
.It Ic os.subplatform
Name of specific operating system variant (e.g.,
.Dq macosx ) .
.br
.Sy Type:
.Em read-only
.It Ic install.user
User for MacPorts installation (e.g.,
.Dq root ) .
.br
.Sy Type:
.Em read-only
.It Ic install.group
Group for MacPorts installation (e.g.,
.Dq wheel ) .
.br
.Sy Type:
.Em read-only
.It Ic applications_dir
Absolute path to the final location to install macOS application bundles
.Pf ( Pa .app
directories).
.br
.Sy Type:
.Em read-only
.br
.Sy Default:
.Em /Applications/MacPorts
.It Ic frameworks_dir
Absolute path to the final location to install macOS framework bundles
.Pf ( Pa .framework
directories).
.br
.Sy Type:
.Em read-only
.br
.Sy Default:
.Em ${prefix}/Library/Frameworks
.El
.Sh DEPENDENCY OPTIONS
Port dependencies should refer to other MacPorts ports
whenever possible, therefore each dependency should be
expressed in the format:
.Bl -tag -width ls -offset indent
.It Li port: Ns Aq port
.El
.Pp
Where
.Aq port
represents the name of an existing MacPorts
.Nm port .
If satisfying a dependency with a MacPorts port is not
practical and it is likely that a dependency must be met
by an Apple optional install, then the alternative dependency
format:
.Bl -tag -width lc -offset indent
.It Ar type Ns Li \&: Ns Ao filename Ac Ns Li \&: Ns Aq port
.El
.Pp
may be used.
Where
.Ar type
is
.Dq bin
if
.Aq filename
is a program,
.Dq lib
if it is a library, or
.Dq path
if it is a path to an installed file.
.br
.Sy Example:
.Dl lib:libX11.6:XFree86
.Bl -tag -width lc
.It Ic depends_fetch
List of dependencies to check before
.Cm fetch ,
.Cm checksum ,
.Cm extract ,
.Cm patch ,
.Cm configure ,
.Cm build ,
.Cm destroot ,
.Cm install ,
and
.Cm package
targets.
.br
.Sy Type:
.Em optional
.br
.Sy Example:
.Dl depends_fetch port:mercurial
.It Ic depends_extract
List of dependencies to check before
.Cm extract ,
.Cm patch ,
.Cm configure ,
.Cm build ,
.Cm destroot ,
.Cm install ,
and
.Cm package
targets.
.br
.Sy Type:
.Em optional
.br
.Sy Example:
.Dl depends_extract path:bin/xz:xz
.It Ic depends_build
List of dependencies to check before
.Cm configure ,
.Cm build ,
.Cm destroot ,
.Cm install ,
and
.Cm package
targets.
.br
.Sy Type:
.Em optional
.br
.Sy Example:
.Dl depends_build port:autoconf
.It Ic depends_run
List of dependencies to check before
.Cm destroot ,
.Cm install
and
.Cm package
targets.
Will be recorded in the registry as being required by the dependent port when
it is installed.
.br
.Sy Type:
.Em optional
.br
.Sy Example:
.Dl depends_run port:bash
.It Ic depends_lib
List of dependencies to check before
.Cm configure ,
.Cm build ,
.Cm destroot ,
.Cm install ,
and
.Cm package
targets.
Will be recorded in the registry as being required by the dependent port when
it is installed.
.br
.Sy Type:
.Em optional
.br
.Sy Example:
.Dl depends_lib port:libfetch
.El
.Sh FETCH OPTIONS
Fetch all distribution files and patches.
.Bl -tag -width lc
.It Ic master_sites.mirror_subdir
Subdirectory to append to all mirror sites for any list specified in
.Ic master_sites .
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em ${name}
.br
.Sy Example:
.Dl master_sites.mirror_subdir      magic
.It Ic patch_sites
List of sites to fetch
.Ic patchfiles
from or a predefined mirror site list.
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em ${master_sites}
.br
.Sy Example:
.Bd -literal -offset indent -compact
patch_sites ftp://ftp.patchcityrepo.com/pub/magic/patches
.Ed
.It Ic patch_sites.mirror_subdir
Subdirectory to append to all mirror sites for any list specified in
.Ic patch_sites .
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em ${name}
.br
.Sy Example:
.Dl patch_sites.mirror_subdir       magic
.It Ic extract.suffix
Suffix to append to
.Ic distname .
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em .tar.gz
.br
.Sy Example:
.Dl extract.suffix .tgz
.It Ic distfiles
List of distribution files to fetch from
.Ic master_sites .
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em ${distname}${extract.suffix}
.br
.Sy Example:
.Dl distfiles magicsource.tar.gz cluebat.tar.bz2
.It Ic patchfiles
List of patches to fetch and apply.
.br
.Sy Type:
.Em optional
.br
.Sy Example:
.Dl patchfiles patch-Makefile.in.diff patch-main.c.diff
.It Ic use_zip
Use zip.
.br
Sets
.Ic extract.suffix
to
.Dq .zip .
.br
Sets
.Ic extract.cmd
to
.Dq unzip .
.br
Sets
.Ic extract.pre_args
to
.Dq -q .
.br
Sets
.Ic extract.post_args
to
.Dq -d ${workpath} .
.br
.Sy Type:
.Em optional
.br
.Sy Example:
.Dl use_zip yes
.It Ic use_bzip2
Use bzip2.
.br
Sets
.Ic extract.suffix
to
.Dq .bz2 .
.br
Sets
.Ic extract.cmd
to
.Dq bzip2 .
.br
.Sy Type:
.Em optional
.br
.Sy Example:
.Dl use_bzip2 yes
.It Ic use_lzip
Use lzip.
.br
Sets
.Ic extract.suffix
to
.Dq .tar.lz .
.br
Sets
.Ic extract.cmd
to
.Dq lzip .
.br
.Sy Type:
.Em optional
.br
.Sy Example:
.Dl use_lzip yes
.It Ic use_lzma
Use lzma.
.br
Sets
.Ic extract.suffix
to
.Dq .lzma .
.br
Sets
.Ic extract.cmd
to
.Dq lzma .
.br
.Sy Type:
.Em optional
.br
.Sy Example:
.Dl use_lzma yes
.It Ic use_tar
Use an uncompressed tar archive.
.br
Sets
.Ic extract.suffix
to
.Dq .tar .
.br
Sets
.Ic extract.cmd
to
.Dq tar .
.br
Sets
.Ic extract.pre_args
to
.Dq -xf .
.br
.Sy Type:
.Em optional
.br
.Sy Example:
.Dl use_tar yes
.It Ic use_xz
Use xz.
.br
Sets
.Ic extract.suffix
to
.Dq .xz .
.br
Sets
.Ic extract.cmd
to
.Dq xz .
.br
.Sy Type:
.Em optional
.br
.Sy Example:
.Dl use_xz yes
.It Ic use_7z
Use 7z (7zip).
.br
Sets
.Ic extract.suffix
to
.Dq .7z .
.br
Sets
.Ic extract.cmd
to
.Dq 7za .
.br
.Sy Type:
.Em optional
.br
.Sy Example:
.Dl use_7z yes
.It Ic use_dmg
Use a disk image (dmg).
Note that the arguments passed to 'hdiutil' are extensive, and can vary based
on the Darwin release, other options, etc.
As such, ports should not attempt to manipulate 'extract.post_args'.
.br
Sets
.Ic extract.suffix
to
.Dq .dmg .
.br
Sets
.Ic extract.cmd
to
.Dq hdiutil .
.br
Sets
.Ic extract.pre_args
to
.Dq attach .
.br
Sets
.Ic extract.post_args
to
.Dq -private -readonly -nobrowse -mountpoint ...etc... .
.br
.Sy Type:
.Em optional
.br
.Sy Example:
.Dl use_dmg yes
.It Ic dist_subdir
Create a sub-directory in
.Ic distpath
to store all fetched files.
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em ${name}
.br
.Sy Example:
.Dl dist_subdir vim${version}
.El
.Ss ADVANCED FETCH OPTIONS
Some mirrors require special options for a resource to be properly
fetched.
.Bl -tag -width lc
.It Ic fetch.user
HTTP or FTP user to fetch the resource.
.br
.Sy Type:
.Em optional
.It Ic fetch.password
HTTP or FTP password to fetch the resource.
.br
.Sy Type:
.Em optional
.It Ic fetch.use_epsv
Whether to use
.Li EPSV
command for FTP transfers.
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em yes
.br
.It Ic fetch.ignore_sslcert
Whether to ignore the host SSL certificate (for HTTPS).
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em no
.br
.El
.Ss FETCHING FROM CVS
As an alternative to fetching distribution files, pulling the sources from a
CVS repository is supported.
Use of CVS can give rise to non-reproducible builds, so it is strongly
discouraged.
.Bl -tag -width lc
.It Ic cvs.root
Specify the address to a CVS repository from which to checkout files.
.br
.Sy Type:
.Em optional
.br
.Sy Default:
.Em none
.br
.Sy Example:
.Bd -literal -offset indent -compact
cvs.root :pserver:anonymous@cvs.sv.gnu.org:/sources/emacs
.Ed
.It Ic cvs.tag
Specify a CVS tag identifying the code to checkout.
.br
.Sy Type:
.Em optional
.br
.Sy Default
.Em none
.br
.Sy Example:
.Dl cvs.tag HEAD
.It Ic cvs.date
A date that identifies the CVS code set to checkout.
.br
.Sy Type:
.Em optional
.br
.Sy Default
.Em none
.br
.Sy Example:
.Dl cvs.date \*q12-April-2005\*q
.It Ic cvs.module
A CVS module from which to check out the code.
.br
.Sy Type:
.Em optional
.br
.Sy Default
.Em none
.br
.Sy Example:
.Dl cvs.module Sources
.El
.Ss FETCHING FROM SUBVERSION
As an alternative to fetching distribution files, pulling the sources from a
subversion repository is supported.
Use of subversion can give rise to non-reproducible builds, so it is strongly
discouraged.
.Bl -tag -width lc
.It Ic svn.url
Specify the url from which to fetch files.
.br
.Sy Type:
.Em required
.br
.Sy Default:
.Em none
.br
.Sy Examples:
.Bd -literal -offset indent -compact
svn.url https://www.example.com/svn/project/tags/1.2.3
svn.url svn://www.example.com/svn/project/tags/1.2.3
.Ed
.It Ic svn.revision
Specify a tag from which svn should fetch files.
This
corresponds to the
.Fl r
option to the svn cli.
Note that you will need to use backslashes to escape characters that have
meaning to the Tcl interpreter, such as braces and double quotes.
.br
.Sy Type:
.Em optional
.br

NEWS
== Overview of changes from MacPorts 2.0.4 to 2.1.0 ==

# For regular users:

*) New 'rev-upgrade' action that checks for and rebuilds ports that may have
   become incorrectly linked when a dependency was upgraded to a new,
   binary-incompatible version. This runs automatically after upgrades and
   installs by default, and can be turned off or configured to only report
   problems in macports.conf.

*) The pkg and mpkg targets will create flat packages on 10.6+.

*) Build-time dependencies will no longer be installed if a prebuilt
   archive is available for download (and port is configured to use it).

*) Config file archive_sites.conf can be used to download archives from
   custom locations.

*) 'buildfromsource' macports.conf option allows permanently choosing
   source-only or binary-only mode.

*) New macports.conf options 'host_blacklist' and 'preferred_hosts', for
   controlling which download hosts will be used.

# For Portfile writers:

*) New -W option for reinplace that sets a base directory for all file
   arguments.

*) Livecheck can now find multiple regex matches on each line.

*) license_noconflict option can be used to override automatic license
   compatibility checks.

*) compiler.whitelist, compiler.blacklist, and compiler.fallback options allow
   simpler, detailed control over which compilers are used to build ports.


== Overview of changes from MacPorts 1.9.2 to 2.0.0 ==

# For regular users:

*) Downloaded archives are now checked for integrity using digital signatures.
   New pubkeys.conf file allows configuring keys to trust.

*) Actions can be abbreviated to an unambiguous prefix of the full name.

*) Ports that create StartupItems (aka launchd plists) should now be
   installable by unprivileged users.

*) Port images are now stored as archives. Archive mode is now effectively
   always on. Direct mode and the flat registry format are no longer
   available. Conversion will happen automatically.

# For Portfile writers:

*) Multiple ports can now be defined in one Portfile using the 'subport'
   procedure.

*) New option 'add_users' can be set to ensure that the listed users are
   created at the right times during the install process.

*) SHA256 checksums are now supported.

== Overview of changes from MacPorts 1.8.2 to 1.9.0 ==

# For regular users:

*) All output is now logged to a file, whose location will be shown if a port
   target fails. This should be attached to bug reports. The new 'log' action
   can also be used to view the log for a port.
*) Command line options can be abbreviated if unambiguous.
*) New actions: rdeps, rdependents, setrequested, unsetrequested
*) --follow-dependencies option for uninstall (ports with the requested flag
   set or with other dependents are skipped).
*) New pseudo-ports and selectors: requested, unrequested, leaves, depof:,
   rdepof:, dependentof:, rdependentof:.
*) Architectures of dependencies are checked before install, and in many cases
   will be automatically fixed by upgrading with +universal.
*) New sqlite registry format which is faster and lacks some of the old
   format's inconvenient limitations.
*) Explicitly deselected variants (e.g. -foo) are now recorded, so they will
   not be set to + when upgrading if +foo appears in variants.conf or
   the port's default_variants.

# For Portfile writers:

*) New os.subplatform variable.
*) New configure.ld_archflags option to match configure.cc_archflags et al.
*) 'platform' acts like an 'if' statement rather than a variant.
*) New supported_archs option for indicating a port only supports certain
   archs or is noarch.
*) pre-/post-deactivate and pre-/post-uninstall procedures can now be used,
   and pre-/post-activate procedures are run every time the port is activated,
   not just the first time when it is installed.

== Overview of changes from MacPorts 1.7.1 to 1.8.0 ==

# For regular users:

*) Compatible with Mac OS X 10.6 Snow Leopard, requires at least 10.4 Tiger
*) CPU architecture to build for is now configurable
*) Performance and memory usage improvements
*) New -y option for dry-run
*) 'port search' has many new options that make it more flexible
*) 'port upgrade' now follows the right dependencies when variants change them
*) 'port info' and 'port variants' indicate selected and default variants;
   'port variants' also shows conflicts and dependencies between variants
*) A port can be marked as being replaced by another port; upgrading will
   perform the replacement
*) upgrade accepts new options: --force, --enforce-variants, --no-replace
*) Ports can be marked as conflicting, and conflicts will be detected before
   the install process starts
*) Proxy information can now be obtained from System Preferences and from
   macports.conf
*) Activating a port will now first deactivate any other version of the same
   port that is already active
*) MacPorts can be configured to drop root privileges when not needed
*) port(1) will exit with an appropriate status code when an error is
   encountered (use -p to continue despite errors)

# For Portfile writers:

*) New options: replaced_by, conflicts, depends_fetch, depends_extract, license
*) build_arch variable indicates which architecture should be targeted. Flags
   used to implement this are configured per-compiler, for example,
   configure.cc_archflags will be passed to configure.cc in CFLAGS
*) Dependencies can be satisfied by ports that are installed but not
   available in the ports tree
*) Deleting all the contents of an option no longer unsets the option entirely
*) Using an alternative fetch or extract type now adds appropriate
   dependencies automatically, e.g. 'use_zip yes' adds a dep on unzip, and
   'fetch.type svn' adds a dep on subversion
*) Running phases prior to install should now work reliably when the port is
   already installed

== Overview of changes from MacPorts 1.6.0 to 1.7.0 ==

# For regular users:


gsoc15-portfile/portfile.el/portfile.el
;;; portfile.el --- Major mode for editing MacPorts Portfiles  -*- lexical-binding: t; -*-

;; Copyright (C) 2015  Chunyang Xu

;; Author: Chunyang Xu <chunyang@macports.org>
;; URL: https://trac.macports.org/browser/branches/gsoc15-portfile/portfile.el/
;; Version: 0.0.1
;; Created: Tue Jun  9 03:14:33 CST 2015
;; Keywords: MacPorts Portfile

;;; Commentary:

;;

;;; Code:

(defvar portfile-required-list
  '("PortSystem" "name" "version" "maintainers"
    "homepage" "platforms"
    "master_sites" "master_sites-append" "master_sites-delete"
    "categories" "categories-append" "categories-delete"
    "description" "description-append"
    "long_description" "long_description-append"))

(defvar portfile-optional-list
  '("PortGroup" "epoch" "revision" "patch_sites"
    "license" "conflicts" "license_noconflict"
    "replaced_by" "supported_archs"
    "distfiles" "distfiles-append" "distfiles-delete"
    "distname" "dist_subdir" "worksrcdir"
    "installs_libs" "nextgroup=PortfileYesNo" "skipwhite"
    "depends_skip_archcheck" "depends_skip_archcheck-append" "depends_skip_archcheck-delete"
    "checksums" "checksums-append" "checksums-delete"))

(defvar portfile-mode-font-lock-keywords
  (list (cons (concat "\\_<"
                      (regexp-opt (append portfile-required-list
                                          portfile-optional-list) t) "\\_>")
              'font-lock-keyword-face)))

(define-derived-mode portfile-mode tcl-mode "portfile-mode"
  "Portfile mode"
  (font-lock-add-keywords nil portfile-mode-font-lock-keywords))

;; Test:
;; (add-hook 'tcl-mode-hook (lambda ()
;;                            (when (string-equal (buffer-name) "Portfile")
;;                              (portfile-mode))))

(provide 'portfile)
;;; portfile.el ends here
